### PR TITLE
[Serialization] Stop pre-loading generic environments.

### PR DESCRIPTION
**Pull request by DougGregor**
Make generic environment deserialization lazy, which eliminates a
significant amount of up-front work. Most clients only need the
generic signature, not the full generic environment.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
That failure is... disturbing and unrelated. @jrose-apple , it's a versioned-objc test failure that's a bit surprising.

**Comment by DougGregor** [comment]
@swift-ci please test compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by jrose-apple** [comment]
Looking into it now.

**Comment by jrose-apple** [comment]
#9663, sorry!

**Comment by DougGregor** [comment]
@swift-ci please smoke test macOS

File: lib/Serialization/Deserialization.cpp [hunk 1]
```diff
@@ -207,12 +207,6 @@ ModuleFile &ModuleFile::getModuleFileForDelayedActions() {
 void ModuleFile::finishPendingActions() {
```
Comment (jrose-apple) [inline lib/Serialization/Deserialization.cpp]: I don't think anyone's using this anymore. Can you remove it in a separate commit?

File: lib/Serialization/ModuleFile.cpp [hunk 2]
```diff
@@ -1297,11 +1297,7 @@ Status ModuleFile::associateWithFileContext(FileUnit *file,
   return getStatus();
 }
 
-ModuleFile::~ModuleFile() {
-  assert(DelayedGenericEnvironments.empty() &&
-         "either finishPendingActions() was never called, or someone forgot "
-         "to use getModuleFileForDelayedActions()");
-}
+ModuleFile::~ModuleFile() { }
```
Comment (jrose-apple) [inline lib/Serialization/ModuleFile.cpp]: Nitpick: `= default`.

**Review by jrose-apple — COMMENTED** [review]
Seems reasonable to me if it all works. Why did we need it in the first place?

File: lib/Serialization/Deserialization.cpp (hunk 1 above)
Comment (DougGregor) [inline lib/Serialization/Deserialization.cpp]: Yeah, sure.

**Comment by DougGregor** [comment]
We needed it because we were getting crashes later on, once the type checker was torn down, because the GSB would cause more protocol conformances to be formed that couldn't be completed. I *think* all of those conformances were ones created by the Clang importer, and that the recent work to make sure the important parts of those (the type witnesses) can be filled in even without a type checker present made these crashes go away. Perhaps you can see why I'm worried about this change :)

**Pull request merged**